File: backend/models/envelope.py
"""Shared API response envelope contracts."""

from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, Field


T = TypeVar("T")


class ApiErrorInfo(BaseModel):
    code: str
    message: str
    details: dict[str, Any] | None = None


class ApiResponseEnvelope(BaseModel, Generic[T]):
    status: Literal["success", "error", "partial"]
    data: T | None = None
    error: ApiErrorInfo | None = None
    meta: dict[str, Any] = Field(default_factory=dict)


class AiQualityGate(BaseModel):
    name: str
    passed: bool
    score: float | None = None
    message: str | None = None


class AiResultEnvelope(BaseModel, Generic[T]):
    status: Literal["success", "partial", "fallback", "error", "low_confidence"]
    data: T | None = None
    fallback: str | None = None
    error: ApiErrorInfo | None = None
    quality_gates: list[AiQualityGate] = Field(default_factory=list)
    trace: dict[str, Any] | None = None
    partial_data: dict[str, Any] | None = None

File: frontend-react/src/services/api/envelope.ts
export type ApiEnvelopeStatus = "success" | "error" | "partial";

export interface ApiErrorInfo {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export interface ApiResponseEnvelope<TData> {
  status: ApiEnvelopeStatus;
  data: TData | null;
  error?: ApiErrorInfo | null;
  meta?: Record<string, unknown>;
}

export type AiResultStatus = "success" | "partial" | "fallback" | "error" | "low_confidence";

export interface AiQualityGate {
  name: string;
  passed: boolean;
  score?: number;
  message?: string;
}

export interface AiResultEnvelope<TData> {
  status: AiResultStatus;
  data: TData | null;
  fallback?: string | null;
  error?: ApiErrorInfo | null;
  quality_gates?: AiQualityGate[];
  trace?: Record<string, unknown> | null;
  partial_data?: Record<string, unknown> | null;
}

export function isApiResponseEnvelope<TData>(value: unknown): value is ApiResponseEnvelope<TData> {
  if (!value || typeof value !== "object") return false;
  const candidate = value as { status?: unknown; data?: unknown };
  return typeof candidate.status === "string" && "data" in candidate;
}

export function unwrapApiResponse<TData>(value: TData | ApiResponseEnvelope<TData>): TData {
  if (isApiResponseEnvelope<TData>(value)) {
    if (value.status === "error") {
      throw new Error(value.error?.message || "API request failed");
    }
    return value.data as TData;
  }
  return value;
}
